src/lib.rs

```diff
@@ -184,388 +121,71 @@ fn validate(payload: &[u8]) -> CallResult {
-                    value: Some("value".to_string()),
-                },
-                settings::EnvVar {
-                    name: "name2".to_string(),
-                    value: Some("value2".to_string()),
-                },
-            ]),
-        };
-        let envvar = vec![
-            apicore::EnvVar {
-                name: "name3".to_string(),
-                value: Some("value3".to_string()),
-                ..Default::default()
-            },
-            apicore::EnvVar {
-                name: "name4".to_string(),
-                value: Some("value4".to_string()),
-                ..Default::default()
-            },
-        ];
-        let result = validate_envvar_with_rule(&rule, &envvar);
-        assert!(result.is_ok());
-        Ok(())
-    }
-
-    #[test]
-    fn anyin_should_reject_if_some_envvar_is_defined_in_the_resource() -> Result<(), ()> {
-        let rule = Rule {
-            reject: settings::Operator::AnyIn,
-            environment_variables: HashSet::from([
-                settings::EnvVar {
-                    name: "name".to_string(),
-                    value: Some("value".to_string()),
-                },
-                settings::EnvVar {
-                    name: "name2".to_string(),
-                    value: Some("value2".to_string()),
-                },
-            ]),
-        };
-        let envvar = vec![
-            apicore::EnvVar {
-                name: "name3".to_string(),
-                value: Some("value3".to_string()),
-                ..Default::default()
-            },
-            apicore::EnvVar {
-                name: "name2".to_string(),
-                value: Some("value2".to_string()),
-                ..Default::default()
-            },
-        ];
-        let result = validate_envvar_with_rule(&rule, &envvar);
-        assert!(result.is_err());
-        assert_eq!(
-            result.unwrap_err().to_string(),
-            format!("{ANY_IN_ERROR_MSG} name2")
+    use rstest::rstest;
+
+    use crate::{
+        CONTAINS_ALL_OF_ERROR_MSG, CONTAINS_ANY_OF_ERROR_MSG, DOES_NOT_CONTAIN_ALL_OF_ERROR_MSG,
+        DOES_NOT_CONTAIN_ANY_OF_ERROR_MSG,
+    };
+
+    #[rstest]
+    #[case::contains_any_of(Settings::ContainsAnyOf{envvars: HashSet::from(["a".to_owned(), "b".to_owned()])}, vec!["c".to_owned()], CONTAINS_ANY_OF_ERROR_MSG)]
+    #[case::does_not_contain_any_of(Settings::DoesNotContainAnyOf{envvars: HashSet::from(["a".to_owned(), "b".to_owned()])}, vec!["b".to_owned(), "c".to_owned()],  DOES_NOT_CONTAIN_ANY_OF_ERROR_MSG)]
+    #[case::contains_all_of(Settings::ContainsAllOf{envvars: HashSet::from(["a".to_owned(), "b".to_owned()])}, vec!["a".to_owned(), "c".to_owned()], CONTAINS_ALL_OF_ERROR_MSG)]
+    #[case::does_not_contain_all_of(Settings::DoesNotContainAllOf{envvars: HashSet::from(["a".to_owned(), "b".to_owned()])}, vec!["a".to_owned(), "b".to_owned()], DOES_NOT_CONTAIN_ALL_OF_ERROR_MSG)]
+    fn test_multiple_container_validation(
```

Comment:
This test is an overlap of the ones inside of `operators.rs`, that's a smell. I think we can restructure the validation logic to make the code itself nicer to read and to test.

I propose to write 3 helper functions:

- `fn get_containers_env_vars(containers &[Container]) -> HashMap<String, Vec<String>>`
- `fn get_ephemeral_containers_env_vars(containers &[EphemeralContainer]) -> HashMap<String, Vec<String>>`

In both cases, the function returns a map with:

- key: the name of the container image
- values: the list of environment variables. Only the names, since the policy does not care about the values

The functions returns and empty dictionary when no env variables are defined.

In this way, we can test these two functions and drop this test.

Reply:
Ok, I didn't realize that was the end goal. This is however being tested inside of operations. Why do we need an extra test covering that?

I was wrongly assuming the final goal of this test was to make sure we're actually looking into the different places of the `PodSpec` that hold container, and hence env variables we need to inspect